libmamba/include/mamba/util/graph.hpp

```diff
@@ -443,127 +456,241 @@ namespace mamba::util
     template <typename UnaryFunc>
     UnaryFunc DiGraphBase<N, G>::for_each_leaf_id_from(node_id source, UnaryFunc func) const
     {
-        struct LeafVisitor : default_visitor<derived_t>
-        {
-            UnaryFunc& m_func;
-
-            LeafVisitor(UnaryFunc& func)
-                : m_func{ func }
-            {
-            }
-
-            void start_node(node_id n, const derived_t& g)
+        dfs_preorder_nodes_for_each_id(
+            derived_cast(),
+            [&](node_id n)
             {
-                if (g.out_degree(n) == 0)
+                if (out_degree(n) == 0)
                 {
-                    m_func(n);
+                    func(n);
                 }
-            }
-        };
-        auto visitor = LeafVisitor(func);
-        depth_first_search(visitor, source);
+            },
+            source
+        );
         return func;
     }
 
     template <typename N, typename G>
     template <typename UnaryFunc>
     UnaryFunc DiGraphBase<N, G>::for_each_root_id_from(node_id source, UnaryFunc func) const
     {
-        struct RootVisitor : default_visitor<derived_t>
-        {
-            UnaryFunc& m_func;
-
-            RootVisitor(UnaryFunc& func)
-                : m_func{ func }
+        dfs_preorder_nodes_for_each_id(
+            derived_cast(),
+            [&](node_id n)
             {
-            }
+                if (in_degree(n) == 0)
+                {
+                    func(n);
+                }
+            },
+            source,
+            /* reverse= */ true
+        );
+        return func;
+    }
+
+    template <typename N, typename G>
+    auto DiGraphBase<N, G>::number_of_node_id() const noexcept -> node_id
```

Comment:
Naming: `successor_count`?

Reply:
No this is meant to be the number of ids generated, for being able to use an `std::vector` instead of a map.

Reply:
Got it